group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -5524,5 +5670,2897 @@ private static Record newGroupMetadataRecord(
+        assertTrue(group.isInState(STABLE));
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void testStaticMemberRejoinWithLeaderIdAndUnknownMemberId(
+        boolean supportSkippingAssignment
+    ) throws Exception {
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .build();
+        GenericGroup group = context.createGenericGroup("group-id");
+
+        RebalanceResult rebalanceResult = staticMembersJoinAndRebalance(
+            context,
+            "group-id",
+            "leader-instance-id",
+            "follower-instance-id"
+        );
+
+        // A static leader rejoin with unknown id will not trigger rebalance, and no assignment will be returned.
+        // As the group was in Stable state and the member id was updated, this will generate records.
+        JoinGroupRequestData joinRequest = new JoinGroupRequestBuilder()
+            .withGroupId("group-id")
+            .withGroupInstanceId("leader-instance-id")
+            .withMemberId(UNKNOWN_MEMBER_ID)
+            .withDefaultProtocolTypeAndProtocols()
+            .build();
+
+        CompletableFuture<JoinGroupResponseData> joinResponseFuture = new CompletableFuture<>();
+        CoordinatorResult<Void, Record> result = context.sendGenericGroupJoin(
+            joinRequest,
+            joinResponseFuture,
+            true,
+            supportSkippingAssignment);
+
+        assertEquals(
+            Collections.singletonList(newGroupMetadataRecord(group, MetadataVersion.latest())),
+            result.records()
+        );
+        // Simulate a successful write to the log.
+        result.appendFuture().complete(null);
+        assertTrue(joinResponseFuture.isDone());
+
+        Set<String> expectedGroupInstanceIds = new HashSet<>();
+        expectedGroupInstanceIds.add("leader-instance-id");
+        expectedGroupInstanceIds.add("follower-instance-id");
+
+        String leader = supportSkippingAssignment ?
+            joinResponseFuture.get().memberId() : rebalanceResult.leaderId;
+
+        List<JoinGroupResponseMember> members = supportSkippingAssignment ?
+            toJoinResponseMembers(group) : Collections.emptyList();
+
+        JoinGroupResponseData expectedJoinResponse = new JoinGroupResponseData()
+            .setErrorCode(Errors.NONE.code())
+            .setGenerationId(rebalanceResult.generationId)
+            .setMemberId(joinResponseFuture.get().memberId())
+            .setLeader(leader)
+            .setProtocolName("range")
+            .setProtocolType("consumer")
+            .setSkipAssignment(supportSkippingAssignment)
+            .setMembers(members);
+
+        checkJoinGroupResponse(
+            expectedJoinResponse,
+            joinResponseFuture.get(),
+            group,
+            STABLE,
+            supportSkippingAssignment ? expectedGroupInstanceIds : Collections.emptySet()
+        );
+    }
```

Comment:
thanks for the catch. I have added them. although
```
    // Calling sync on old leader.id will fail because that leader.id is no longer valid and replaced.
    val newLeaderSyncGroupResult = syncGroupLeader(groupId, rebalanceResult.generation, rebalanceResult.leaderId, Map.empty)
    assertEquals(Errors.UNKNOWN_MEMBER_ID, newLeaderSyncGroupResult.error)
```
is a bit awkward; that request fails with UNKNOWN_MEMBER_ID because the group instance id is missing

Reply:
Have you pushed the code for this? I can't find them.

Reply:
Yes, L6407 - L6438